File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:
        res = []
        path = []

        def backtrack(open_count: int, close_count: int) -> None:
            if len(path) == 2 * n:
                res.append(''.join(path))
                return

            # Can add '(' as long as we haven't used all n opens
            if open_count < n:
                path.append('(')
                backtrack(open_count + 1, close_count)
                path.pop()

            # Can add ')' only if it wouldn't create more closes than opens
            if close_count < open_count:
                path.append(')')
                backtrack(open_count, close_count + 1)
                path.pop()

        backtrack(0, 0)
        return res

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        string path;
        path.reserve(2 * n);

        backtrack(n, 0, 0, path, res);
        return res;
    }

private:
    void backtrack(int n, int openCount, int closeCount, string& path, vector<string>& res) {
        if ((int)path.size() == 2 * n) {
            res.push_back(path);
            return;
        }

        if (openCount < n) {
            path.push_back('(');
            backtrack(n, openCount + 1, closeCount, path, res);
            path.pop_back();
        }

        if (closeCount < openCount) {
            path.push_back(')');
            backtrack(n, openCount, closeCount + 1, path, res);
            path.pop_back();
        }
    }
};